### PR TITLE
Run CI/CD only on Source Code

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
## Summary

Jenkins will now skip ci/cd pipeline if source code file paths defied in the Jenkinsfile do not have changes.

## Checklist

- [x] Math issue #1658 

- [ ] Copyright holder: (fill in copyright holder information)

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -145,6 +161,44 @@ pipeline {
...
+        stage('Verify changes') {
+            agent any
+            steps{
+                script {         
+
+                    retry(3) { checkout scm }
+                    sh "git clean -xffd"
+
+                    def commitHash = sh(script: "git rev-parse HEAD | tr '\\n' ' '", returnStdout: true)
+                    sh(script: "git pull && git checkout develop", returnStdout: false)
```
Comment (rok-cesnovar) [inline Jenkinsfile]: Can Jenkins figure out to which branch a PR is done to?

I tried to run a PR to a non-develop branch and that fails. See https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1725/2/pipeline/

File: Jenkinsfile [hunk 2]
```diff
@@ -25,6 +25,21 @@ def deleteDirWin() {
     deleteDir()
 }
 
+def sourceCodePaths(){
+    // These paths will be passed to git diff
+    // If there are changes to them, CI/CD will continue else skip
+    def paths = ['src', 'make', 'lib', 'test', 'runTests.py', 'runChecks.py', 'makefile', 'Jenkinsfile', '.clang-format']
```
Comment (rok-cesnovar) [inline Jenkinsfile]: The first one should be `stan` not `src`.

File: Jenkinsfile [hunk 3]
```diff
@@ -145,6 +161,44 @@ pipeline {
...
+                    }
+                }
+            }
+            post {
+                always {
+                    deleteDir()
+                }
+            }
+        }
         stage('Headers checks') {
```
Comment (rok-cesnovar) [inline Jenkinsfile]: Can we also not run headers checks if there are no changes? I think there is no reason to run the headers check here.

File: Jenkinsfile (hunk 1 above)
Comment (rok-cesnovar) [reply Jenkinsfile]: If this is not doable or hard to do, I am also fine with skipping tests on PRs not done to develop. They will have to be done to develop eventually anyways.

Comment (serban-nicusor-toptal) [reply Jenkinsfile]: Hey, based on the [environment variables](https://ci.eclipse.org/webtools/env-vars.html/) we can use `CHANGE_TARGET` here, I've pushed a change now.

Comment (rok-cesnovar) [reply Jenkinsfile]: Great! This all works now https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1725/14/pipeline

Comment (serban-nicusor-toptal) [reply Jenkinsfile]: sweet!

**Comment by mcol** [comment]
Just to confirm: would the upload doxygen step be skipped with this or not? I think we would like it to happen, right?

**Comment by serban-nicusor-toptal** [comment]
Hey @mcol ,

The Upload Doxygen was defined to happen only on develop, see [here](https://github.com/stan-dev/math/blob/skip-nonsc-cicd/Jenkinsfile#L384)
Let me know if that should be changed.

**Comment by rok-cesnovar** [comment]
It would not be skipped as that stage stayed the same. This wasnt tested because it only happens on develop.

**Comment by mcol** [comment]
OK, thanks I couldn't figure it out looking at the patch or at the Jenkins logs. :)

**Comment by mcol** [comment]
The upstream test have failed, hopefully it's unrelated.

**Comment by rok-cesnovar** [comment]
Its not related to this but the cmdstan PR for the same CI source code thing https://github.com/stan-dev/cmdstan/pull/824

Something is different regarding tests on develop I guess.

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal  do we need the same fix as on cmdstan here?

**Comment by serban-nicusor-toptal** [comment]
Yes and I found something else, I will update this PR after a fix on cmdstan.

**Comment by rok-cesnovar** [comment]
This works great! One more minor left to do is to also skip the merge tests. It isnt a big deal but since we are already doing this.
See https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/develop/568/pipeline

**Comment by serban-nicusor-toptal** [comment]
Should we skip `Upload doxygen` too ?

**Comment by rok-cesnovar** [comment]
No, doxygen changes should get pushed.

**Pull request merged**